golang/trace/zipkin_tracereporter.go

```diff
@@ -0,0 +1,116 @@
+package trace
+
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"net"
+	"time"
+
+	tc "github.com/uber/tchannel/golang"
+	"github.com/uber/tchannel/golang/thrift"
+	"github.com/uber/tchannel/golang/trace/thrift/gen-go/tcollector"
+)
+
+// ZipkinTraceReporter is a trace reporter that submits trace spans in to zipkin trace server.
+type ZipkinTraceReporter struct {
+	tchannel *tc.Channel
+	client   tcollector.TChanTCollector
+}
+
+// NewZipkinTraceReporter returns a zipkin trace reporter that submits span to tcollector service.
+func NewZipkinTraceReporter(ch *tc.Channel) *ZipkinTraceReporter {
+	thriftClient := thrift.NewClient(ch, "tcollector", nil)
+	client := tcollector.NewTChanTCollectorClient(thriftClient)
+	return &ZipkinTraceReporter{tchannel: ch, client: client}
+}
+
+// Report method will submit trace span to tcollector server.
+func (r *ZipkinTraceReporter) Report(span tc.Span, annotations []tc.Annotation, binaryAnnotations []tc.BinaryAnnotation) {
+	ctx, cancel := tc.NewContextBuilder(time.Second).
+		SetShardKey(base64Encode(span.TraceID())).Build()
+	defer cancel()
+
+	// FIXME remove this dummy endpoint.
+	endpoint := &tc.Endpoint{Ipv4: "127.0.0.1", Port: 8888, ServiceName: "dummy"}
+	thriftSpan := buildZipkinSpan(span, annotations, binaryAnnotations, "dummy", endpoint)
+	// client submit
+	r.client.Submit(ctx, thriftSpan)
+}
+
+// buildZipkinSpan builds zipkin span based on tchannel span.
+func buildZipkinSpan(span tc.Span, annotations []tc.Annotation, binaryAnnotations []tc.BinaryAnnotation, name string, endpoint *tc.Endpoint) *tcollector.Span {
+	host := tcollector.Endpoint{
+		Ipv4:        (int32)(inetAton(endpoint.Ipv4)),
+		Port:        endpoint.Port,
+		ServiceName: endpoint.ServiceName,
+	}
+
+	// TODO Add BinaryAnnotations
+	thriftSpan := tcollector.Span{
+		TraceId:     uInt64ToBytes(span.TraceID()),
+		Host:        &host,
+		Name:        name,
+		Id:          uInt64ToBytes(span.SpanID()),
+		ParentId:    uInt64ToBytes(span.ParentID()),
+		Annotations: buildZipkinAnnotations(annotations),
+		Debug:       false,
+	}
+
+	return &thriftSpan
+}
+
+// buildZipkinAnnotations builds zipkin Annotations based on tchannel annotations.
+func buildZipkinAnnotations(anns []tc.Annotation) []*tcollector.Annotation {
+	zipkinAnns := make([]*tcollector.Annotation, len(anns))
+	for i, ann := range anns {
+		zipkinAnns[i] = &tcollector.Annotation{
+			Timestamp: (float64)(ann.Timestamp.UnixNano() / 1e6),
+			Value:     (string)(ann.Key),
+		}
+	}
+	return zipkinAnns
+}
+
+// inetAton converts string Ipv4 to uint32
+func inetAton(ip string) uint32 {
+	ipBytes := net.ParseIP(ip).To4()
+	return binary.BigEndian.Uint32(ipBytes)
+}
+
+// base64Encode encodes uint64 with base64 StdEncoding.
+func base64Encode(data uint64) string {
+	return base64.StdEncoding.EncodeToString(uInt64ToBytes(data))
+}
+
+// UInt64ToBytes converts uint64 to bytes.
+func uInt64ToBytes(i uint64) []byte {
```

Comment:
This should be called `uint64ToBytes` (and make sure the comment is updated to reflect the new name)